source/Octopus.Server.Client/Operations/RegisterMachineOperation.cs

```diff
@@ -156,31 +158,10 @@ List<EnvironmentResource> GetEnvironments(IOctopusSpaceRepository repository)
 
             if (Environments is not null && Environments.Any())
             {
-                var envsByName = repository.Environments.FindByNames(EnvironmentNames);
-                environments.AddRange(envsByName);
-
-                var missing = Environments
-                    .Except(envsByName.Select(e => e.Name), StringComparer.OrdinalIgnoreCase)
-                    .ToArray();
-
-                //use the missing names to try and find by slug
-                var environmentsBySlug = repository.Environments.FindBySlugs(missing);
-                environments.AddRange(environmentsBySlug);
-
-                missing = missing
-                    .Except(environmentsBySlug.Select(e => e.Slug), StringComparer.OrdinalIgnoreCase)
-                    .ToArray();
-
-                //any other missing slugs/names could be Id's, so looks again
-                var environmentByIds = repository.Environments.Get(missing);
-                environments.AddRange(environmentByIds);
-
-                missing = missing
-                    .Except(environmentByIds.Select(e => e.Id), StringComparer.OrdinalIgnoreCase)
-                    .ToArray();
-
-                if (missing.Any())
-                    throw new InvalidRegistrationArgumentsException(CouldNotFindByMultipleMessage("environment", missing.ToArray()));
+                var environmentsByNameIdOrSlug =
+                    repository.Environments.FindByNameIdOrSlugs<EnvironmentResource, IEnvironmentRepository>(
```

Comment:
Unfortunately this is because the caller needs to provide a custom error message that contains the type of resource it couldn't find (I think this is because it needs to be in the error logs in Tentacle.. and Client.Operations is largely concerned about Tentacle). In any case, the actual error message generator comes from a protected static method defined in `RegisterMachineOperationBase`, presumably specific to machine operations. Using a lambda to preserve existing error logging was the best solution I could come up with. I didn't want the generic extension method to start knowing about machine operation error messages.